playground/src/main/kotlin/com/strumenta/kolasu/playground/EMFSupport.kt

```diff
@@ -0,0 +1,140 @@
+        addAttribute("path", string, 1, 1)
+        addAttribute("code", string, 1, 1)
+        addContainment("result", result, 1, 1)
+    }
+
+    ePackage.createEClass("WorkspaceTranspilationTrace").apply {
+        addContainment("originalFiles", workspaceFile, 1, -1)
+        addContainment("generatedFiles", workspaceFile, 1, -1)
+        addContainment("issues", issue, 0, -1)
+    }
+
+    return ePackage
+}
+
+private fun makeTranspilationTraceEObject(): EObject {
+    val ec = TRANSPILATION_METAMODEL.getEClass(TranspilationTrace::class.java)
+    return TRANSPILATION_METAMODEL.eFactoryInstance.create(ec)
+}
+
+private fun makeWorkspaceTranspilationTraceEObject(): EObject {
+    val ec = TRANSPILATION_METAMODEL.getEClass(WorkspaceTranspilationTrace::class.java)
+    return TRANSPILATION_METAMODEL.eFactoryInstance.create(ec)
+}
+
+private fun makeWorkspaceFileEObject(): EObject {
+    val ec = TRANSPILATION_METAMODEL.getEClass("WorkspaceFile")
+    return TRANSPILATION_METAMODEL.eFactoryInstance.create(ec)
+}
+
+fun <S : Node, T : Node> TranspilationTrace<S, T>.toEObject(resource: Resource): EObject {
+    val transpilationTraceEO = makeTranspilationTraceEObject()
+    transpilationTraceEO.setStringAttribute("originalCode", this.originalCode)
+    val mapping = KolasuToEMFMapping()
+    transpilationTraceEO.setSingleContainment("sourceResult", this.sourceResult.toEObject(resource, mapping))
+    transpilationTraceEO.setSingleContainment("targetResult", this.targetResult.toEObject(resource, mapping))
+    transpilationTraceEO.setStringAttribute("generatedCode", this.generatedCode)
+    transpilationTraceEO.setMultipleContainment("issues", this.transpilationIssues.map { it.toEObject() })
+    return transpilationTraceEO
+}
+
+fun WorkspaceTranspilationTrace.WorkspaceFile.toEObject(resource: Resource, mapping: KolasuToEMFMapping): EObject {
+    val workspaceFileEO = makeWorkspaceFileEObject()
+    workspaceFileEO.setStringAttribute("path", this.path)
+    workspaceFileEO.setStringAttribute("code", this.code)
+    workspaceFileEO.setSingleContainment("result", this.result.toEObject(resource, mapping))
+    return workspaceFileEO
+}
+
+fun WorkspaceTranspilationTrace.toEObject(resource: Resource): EObject {
+    val transpilationTraceEO = makeWorkspaceTranspilationTraceEObject()
+    val mapping = KolasuToEMFMapping()
+
+    transpilationTraceEO.setMultipleContainment(
+        "originalFiles",
+        this.originalFiles.map { it.toEObject(resource, mapping) }
+    )
+    transpilationTraceEO.setMultipleContainment(
+        "generatedFiles",
+        this.generatedFiles.map { it.toEObject(resource, mapping) }
+    )
+
+    transpilationTraceEO.setMultipleContainment("issues", this.transpilationIssues.map { it.toEObject() })
+    return transpilationTraceEO
+}
+
+fun <S : Node, T : Node> TranspilationTrace<S, T>.saveAsJson(name: String, vararg ePackages: EPackage): String {
```

Comment:
Don't we have generic methods to do this already?

Reply:
You are right. I think we are doing that just to invoke our own methods to convert certain classes to EObject, such as:
```
fun <S : Node, T : Node> TranspilationTrace<S, T>.toEObject(resource: Resource): EObject {
    val transpilationTraceEO = makeTranspilationTraceEObject()
    transpilationTraceEO.setStringAttribute("originalCode", this.originalCode)
    val mapping = KolasuToEMFMapping()
    transpilationTraceEO.setSingleContainment("sourceResult", this.sourceResult.toEObject(resource, mapping))
    transpilationTraceEO.setSingleContainment("targetResult", this.targetResult.toEObject(resource, mapping))
    transpilationTraceEO.setStringAttribute("generatedCode", this.generatedCode)
    transpilationTraceEO.setMultipleContainment("issues", this.transpilationIssues.map { it.toEObject() })
    return transpilationTraceEO
}

fun WorkspaceTranspilationTrace.WorkspaceFile.toEObject(resource: Resource, mapping: KolasuToEMFMapping): EObject {
    val workspaceFileEO = makeWorkspaceFileEObject()
    workspaceFileEO.setStringAttribute("path", this.path)
    workspaceFileEO.setStringAttribute("code", this.code)
    workspaceFileEO.setSingleContainment("result", this.result.toEObject(resource, mapping))
    return workspaceFileEO
}
```

We could probably defining some system to define serializers/unserializers and avoiding this duplication. Maybe we can open an issue for that, what do you think?

Reply:
Sure, makes sense

Reply:
I created this issue: https://github.com/Strumenta/kolasu/issues/230